### PR TITLE
feat(plugin-chart-echarts): support non-timeseries x-axis

### LINKED ISSUE #6568: Make line chart support non-temporal variable for x-axis
The x-axis need not necessarily be a time/date, it could just be any number. Would be great to make that possible!

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
A first step in the direction of adding support for non-temporal x-axes to ECharts line/area/bar/step/scatter charts. When the feature flag is enabled, a new control "X-Axis" shows up which accepts any type of column, be it temporal, numerical or categorical. For new charts the control is mandatory, but old (temporal) charts will still work. Other changes:
- Remove the "Time-series" prefix from all ECharts timeseries charts when 
- make formula annotations work with numeric x-axes
- clean up some TS types/interfaces that had fallen out of sync with the post processing API in the backend

### SCREENSHOTS
When the feature flag is enabled, the chart picker now shows all old ECharts time-series charts without the prefix (the descriptions are also updated to reflect that the chart is now "generic"):
image

Bar chart v2 (previously Time-series bar chart) now works with non-temporal x-axis:
image

Another example; Area chart v2, with the x-axis labels rotated 45 degrees (previously Time-series area chart):
image

### VIDEO
https://user-images.githubusercontent.com/33317356/148947353-9a027210-a9e4-42ab-93b5-bf356261f359.mp4

### TESTING INSTRUCTIONS
1. Enable the `GENERIC_CHART_AXES` feature flag
2. Open the chart picker and notice all "Time-series" charts have been replaced with a more generic name "Time-series line chart" - > "Line chart"
3. Drop a categorical column in the "x axis" control
4. Notice that a line chart is created with a non-temporal x-axis

### ADDITIONAL INFORMATION

- [x] Has associated issue: closes #6568
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true FEATURE_GENERIC_CHART_AXES=true

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true FEATURE_GENERIC_CHART_AXES=true

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true FEATURE_GENERIC_CHART_AXES=true

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/EchartsTimeseries.tsx [hunk 1]
```diff
@@ -130,7 +130,9 @@ export default function EchartsTimeseries({
       }
       // Ensure that double-click events do not trigger single click event. So we put it in the timer.
       clickTimer.current = setTimeout(() => {
-        const { seriesName: name } = props;
+        const { seriesName: name, value } = props;
+        const xValue = value[0].getTime?.() || value[0];
+        console.log(xValue);
```
Comment (stephenLYZ) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/EchartsTimeseries.tsx:135]: we can remove it if this log is not needed

Comment (villebro) [reply superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/EchartsTimeseries.tsx:135]: Oops!

File: superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Area/index.ts [hunk 2]
```diff
@@ -63,7 +55,7 @@ export default class EchartsAreaChartPlugin extends ChartPlugin<
         category: t('Evolution'),
         credits: ['https://echarts.apache.org'],
         description: t(
-          'Time-series Area chart are similar to line chart in that they represent variables with the same scale, but area charts stack the metrics on top of each other. An area chart in Superset can be stream, stack, or expand.',
+          'Area charts are similar to line charts in that they represent variables with the same scale, but area charts stack the metrics on top of each other.',
```
Comment (kgabryje) [inline superset-frontend/plugins/plugin-chart-echarts/src/Timeseries/Area/index.ts:58]: We probably should change that description only if FF is enabled, like in other charts

**Review by kgabryje — CHANGES_REQUESTED** [review]
When user leaves X-Axis control empty while using a dataset without dttm column, an error pops up. I think we should either make x axis field required or update the error message.

image

That however is the only problem that I could find - otherwise this PR is absolutely fantastic!

**Review by zhaoyongjie — COMMENTED** [review]
This is an important feature for an enterprise-level BI, thank you for your work.

I pulled PR to the local and make a basic function test.

1. do not apply time format to category axis
image

**Comment by zhaoyongjie** [comment]
2. I apply a string axis and checked `stack series` and `only total`, then the bar value show `NaNundefined`
image

**Comment by signals2** [comment]
what a big step... when will this feature get ready for test? I've been waiting for this feature for several years

**Comment by rumbin** [comment]
@villebro Thanks a lot for working in this feature!
This will help us to gain more users in our company, who are currently using Excel or PowerBI because of the missing X/Y charts.

**Comment by schickler-c-mayer** [comment]
This is a HUGE step forward for Superset to step out of the data analytics niche and move into the coporate world. We are using Superset heavily in data analytics projects, but haven't been able to bring it to our clients world yet. This feature will change this quite a bit. Thank you!

**Comment by villebro** [comment]
> This is a HUGE step forward for Superset to step out of the data analytics niche and move into the coporate world. We are using Superset heavily in data analytics projects, but haven't been able to bring it to our clients world yet. This feature will change this quite a bit. Thank you!

Thanks so much for the encouraging words @schickler-c-mayer! Hearing feedback like this really means a lot! ❤️

**Comment by villebro** [comment]
> what a big step... when will this feature get ready for test? I've been waiting for this feature for several years

Awesome to hear! ❤️ Well, I hope it will be merged fairly soon, so after that feel free to flip the feature flag and do some testing and report any issues you run into! 🚀

**Comment by villebro** [comment]
@kgabryje and @zhaoyongjie - thanks so much for the reviews and testing! ❤️  I've now addressed the main issues you identified:
- The x-axis control is now mandatory when the feature flag has been flipped (charts with the old metadata that's missing the x-axis control value will still work as before on dashboards). I anticipate this will be the new flow going forward, so IMO this change makes sense. https://github.com/apache/superset/pull/17917/commits/b580e1e83ef4cad6e2e2e63c4d62d79e30e9371f
- The x-axis formatter problem has now been fixed: https://github.com/apache/superset/pull/17917/commits/c6834a8395b8d8b86b61e29b377cbba54a0d6a98
- The "Show values" bug has also been fixed: https://github.com/apache/superset/pull/17917/commits/61418394c46ed5be026b3fe696f5d8fe208843a5

In other words, this should be ready for another round of reviews/testing (I'll restart the eph env once CI is ready).

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true FEATURE_GENERIC_CHART_AXES=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by kgabryje** [comment]
Advanced analytics don't work when a temporal column is selected as X Axis:
image

Current results for reference:
image

**Review by zhaoyongjie — APPROVED** [review]
Thanks for your hard work! This is an amazing function!

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true FEATURE_GENERIC_CHART_AXES=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by kgabryje** [comment]
@villebro When I added Time Comparison in Advanced Analytics, I got a db engine error

image

**Comment by kgabryje** [comment]
And something similar when I use Resample

image

**Comment by villebro** [comment]
Thanks again for the testing help @kgabryje and @zhaoyongjie ! Here's a recap of the latest updates:
1. the implicit operators have been replaced in https://github.com/apache/superset/pull/17917/commits/cf49e33630d82f9e3dfa7a7d1d088fe015424a9c and https://github.com/apache/superset/pull/17917/commits/934165d435973d8725f06a3a6320f63ae2569578
2. I removed the `TIME_COLUMN` constant from `@superset-ui/chart-controls`, as it was a duplicate of the more commonly used `DTTM_ALIAS` in `@superset-ui/core`: https://github.com/apache/superset/pull/17917/commits/8b7a6381e386de08420a5fe2b1c5f9becb986692

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_DASHBOARD_NATIVE_FILTERS_SET=true FEATURE_DASHBOARD_FILTERS_EXPERIMENTAL=true FEATURE_GENERIC_CHART_AXES=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by villebro** [comment]
FYI, some Advanced Analytics features (Prophet, Time shift) won't fully work on Sqlite due to that database returning temporal data as strings. Getting around that needs a PR of its own to avoid bloating this PR

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_ENABLE_DND_WITH_CLICK_UX=true FEATURE_UX_BETA=true

**Comment by kgabryje** [comment]
Oh and 1 more nit - currently we only support time formatting for x axis, we probably should add number formatting as well (might be a task for future PR though)

image

**Comment by villebro** [comment]
@kgabryje - to keep this PR from becoming more complex, I suggest merging this as-is once QA gives a green light and then tackling the remaining known issues in follow up PRs. I'm going to start a project board to track known tasks so we can move this work forward as quickly as possible and make it possible for as many people to help out as possible.

**Comment by junlinccc** [comment]
❤️

**Review by kgabryje — APPROVED** [review]
That sounds good, approved!

**Comment by schickler-c-mayer** [comment]
@villebro 🙏🏻 🙇‍♂️ ❤️

**Comment by signals2** [comment]
@villebro my hero, is this feature ready to merge? I could hardly wait for another day.

**Comment by villebro** [comment]
> villebro my hero, is this feature ready to merge? I could hardly wait for another day.

@signals2 we're still waiting for some more testing from our QA team to ensure this PR doesn't introduce any regressions. But don't fear, I'm sure it will be merged soon 👍

**Comment by jinghua-qa** [comment]
Hi I have found 2 issue when FF is off:

Issue 1, when i add group by, it is not grouping the data. 
Repo steps: use clean_sales_data, select times bar chart v2
- use metric sum(sales)
- add group by year
- observe the chart
Screen Shot 2022-01-14 at 8 41 08 AM

This is what i am expected: (result from current master)
Screen Shot 2022-01-14 at 8 42 39 AM

**Comment by villebro** [comment]
With feature flag enabled:
image

With feature flag disabled:
image

**Comment by kazzasticc** [comment]
> ### SUMMARY
> A first step in the direction of adding support for non-temporal x-axes to ECharts line/area/bar/step/scatter charts. When the feature flag is enabled, a new control "X-Axis" shows up which accepts any type of column, be it temporal, numerical or categorical. For new charts the control is mandatory, but old (temporal) charts will still work. Other changes:
> 
> * Remove the "Time-series" prefix from all ECharts timeseries charts when
> * make formula annotations work with numeric x-axes
> * clean up some TS types/interfaces that had fallen out of sync with the post processing API in the backend
> 
> ### SCREENSHOTS
> When the feature flag is enabled, the chart picker now shows all old ECharts time-series charts without the prefix (the descriptions are also updated to reflect that the chart is now "generic"): image
> 
> Bar chart v2 (previously Time-series bar chart) now works with non-temporal x-axis: image
> 
> Another example; Area chart v2, with the x-axis labels rotated 45 degrees (previously Time-series area chart): image
> 
> ### VIDEO
>  non-temp-x.mp4 
> ### TESTING INSTRUCTIONS
> 1. Enable the `GENERIC_CHART_AXES` feature flag
> 2. Open the chart picker and notice all "Time-series" charts have been replaced with a more generic name "Time-series line chart" - > "Line chart"
> 3. Drop a categorical column in the "x axis" control
> 4. Notice that a line chart is created with a non-temporal x-axis
> 
> ### ADDITIONAL INFORMATION
> * [x]  Has associated issue: closes [Make line chart support non-temporal variable for x-axis #6568](https://github.com/apache/superset/issues/6568)
> * [ ]  Required feature flags:
> * [ ]  Changes UI
> * [ ]  Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
>   
>   * [ ]  Migration is atomic, supports rollback & is backwards-compatible
>   * [ ]  Confirm DB migration upgrade and downgrade tested
>   * [ ]  Runtime estimates and downtime expectations provided
> * [ ]  Introduces new feature or API
> * [ ]  Removes existing feature or API

Has the GROUP BY field been removed?
